cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java

```diff
@@ -0,0 +1,466 @@
+    } finally {
+      partitions.close();
+    }
+  }
+
+  @Override
+  public <T> Class<? extends T> getInputFormatClass() {
+    return files.getInputFormatClass();
+  }
+
+  @Override
+  public Map<String, String> getInputFormatConfiguration() {
+    PartitionFilter filter;
+    try {
+      filter = PartitionedFileSetArguments.getInputPartitionFilter(runtimeArguments, partitioning);
+    } catch (Exception e) {
+      throw new DataSetException("Partition filter must be correctly specified in arguments.");
+    }
+    Collection<String> inputPaths = getPartitionPaths(filter);
+    List<Location> inputLocations = Lists.newArrayListWithExpectedSize(inputPaths.size());
+    for (String path : inputPaths) {
+      inputLocations.add(files.getLocation(path));
+    }
+    return files.getInputFormatConfiguration(inputLocations);
+  }
+
+  @Override
+  public <T> Class<? extends T> getOutputFormatClass() {
+    // we verify that the output partition time is configured in getOutputFormatConfiguration()
+    // todo use a wrapper that adds the new partition when the job is committed - that means we must serialize the
+    //      cconf into the hadoop conf to be able to instantiate this dataset in the output committer
+    return files.getOutputFormatClass();
+  }
+
+  @Override
+  public Map<String, String> getOutputFormatConfiguration() {
+    // all runtime arguments are passed on to the file set, so we can expect the partition time in the file set's
+    // output format configuration. If it is not there, the output format will fail to register this partition.
+    Map<String, String> config = files.getOutputFormatConfiguration();
+    Long time = TimePartitionedFileSetArguments.getOutputPartitionTime(runtimeArguments);
+    if (time == null) {
+      throw new DataSetException("Time must be given for the new output partition as a runtime argument.");
+    }
+    // add the output partition time to the output arguments of the embedded file set
+    Map<String, String> outputArgs = Maps.newHashMap();
+    outputArgs.putAll(config);
+    TimePartitionedFileSetArguments.setOutputPartitionTime(outputArgs, time);
+    return ImmutableMap.copyOf(outputArgs);
+  }
+
+  @Override
+  public FileSet getEmbeddedFileSet() {
+    return files;
+  }
+
+  @Override
+  public Map<String, String> getRuntimeArguments() {
+    return runtimeArguments;
+  }
+
+  //------ private helpers below here --------------------------------------------------------------
+
+  @VisibleForTesting
+  static byte[] generateRowKey(PartitionKey key, Partitioning partitioning) {
+    // validate partition key, convert values, and compute size of output
+    Map<String, FieldType> partitionFields = partitioning.getFields();
+    int totalSize = partitionFields.size() - 1; // one \0 between each of the fields
+    ArrayList<byte[]> values = Lists.newArrayListWithCapacity(partitionFields.size());
+    for (Map.Entry<String, FieldType> entry : partitionFields.entrySet()) {
```

Comment:
So this requires the `Partitioning.getFields()` returns a map with order preserved, right? I known from the implementation of Partitioning it is, however the `getFields` method javadoc doesn't mention it. I think it's good to have it as the `getFields` method contract so that it won't get messed up in future.